Firestore/core/src/firebase/firestore/api/query_core.mm

```diff
@@ -273,25 +267,47 @@ QuerySnapshot result(firestore_, query_, std::move(snapshot),
   return Wrap([query() queryByAddingEndAt:bound]);
 }
 
-void Query::ValidateNewRelationFilter(const RelationFilter& filter) const {
-  if (filter.IsInequality()) {
-    const FieldPath* existing_field = [query_ inequalityFilterField];
-    if (existing_field && *existing_field != filter.field()) {
-      ThrowInvalidArgument(
-          "Invalid Query. All where filters with an inequality (lessThan, "
-          "lessThanOrEqual, greaterThan, or greaterThanOrEqual) must be on the "
-          "same field. But you have inequality filters on '%s' and '%s'",
-          existing_field->CanonicalString(), filter.field().CanonicalString());
-    }
+namespace {
 
-    const FieldPath* first_order_by_field = [query_ firstSortOrderField];
-    if (first_order_by_field) {
-      ValidateOrderByField(*first_order_by_field, filter.field());
-    }
-  } else if (filter.op() == Filter::Operator::ArrayContains) {
-    if ([query_ hasArrayContainsFilter]) {
-      ThrowInvalidArgument(
-          "Invalid Query. Queries only support a single arrayContains filter.");
+constexpr Operator kArrayOps[] = {
+    Operator::ArrayContains,
+};
+
+}
+
+void Query::ValidateNewFilter(const class Filter& filter) const {
+  if (filter.IsFieldFilter()) {
+    const auto& field_filter = static_cast<const FieldFilter&>(filter);
+
+    if (field_filter.IsInequality()) {
+      const FieldPath* existing_inequality = [query_ inequalityFilterField];
+      const FieldPath* new_inequality = &filter.field();
+
+      if (existing_inequality && *existing_inequality != *new_inequality) {
+        ThrowInvalidArgument(
+            "Invalid Query. All where filters with an inequality (lessThan, "
+            "lessThanOrEqual, greaterThan, or greaterThanOrEqual) must be on "
+            "the "
+            "same field. But you have inequality filters on '%s' and '%s'",
+            existing_inequality->CanonicalString(),
+            new_inequality->CanonicalString());
+      }
+
+      const FieldPath* first_order_by_field = [query_ firstSortOrderField];
+      if (first_order_by_field) {
+        ValidateOrderByField(*first_order_by_field, filter.field());
+      }
+
+    } else {
+      // You can have at most 1 disjunctive filter and 1 array filter. Check if
+      // the new filter conflicts with an existing one.
+      Operator filter_op = field_filter.op();
```

Comment:
Optional: this variable doesn't seem to add much value.

Reply:
The next change that Brian adds makes further use of it.